lib/hanami/slice.rb

```diff
@@ -270,6 +281,15 @@ def prepare_all
         prepare_container_providers
       end
```

Comment:
I really think this should be made abstract at some point post 2.0. We could have something like boot/preparation stages and a public API to add your own. It should also be possible to define dependencies between stages so that we can be sure things are run in the corrected order. Proper exception handling with meaningful messages could be achieved too as a nice side-effect, so that IF your app crashes during boot, you don't get an obscure stack trace and instead you clearly see what happened (with a trace of stage execution and the final error message).

Personally, I think Hanami boot/preparation process shouldn't be hardcoded and/or hidden as some kind of a special magical process that happens behind the scenes. It should be public and documented as it's a fundamental piece of the stack. This is another thing that would make DX just nicer.

I can write an issue about it and add to trello if you think this would make sense.

Reply:
@solnic 100% agree. I've always considered this manually-managed app boot to be a compromise we have to make in order to get something shipped for 2.0.

For a subsequent release I'd love to turn it into something more generic, and something that provides hooks for users or third-party gem providers to add their own boot logic into.